mesonbuild/interpreter/interpreter.py

```diff
@@ -1630,28 +1630,10 @@ def find_program_impl(self, args: T.List[mesonlib.FileOrString],
                 raise InterpreterException(m.format(progobj.get_name()))
             return progobj
 
-        if wanted:
-            if version_func:
-                version = version_func(progobj)
-            elif isinstance(progobj, build.Executable):
-                if progobj.subproject:
-                    interp = self.subprojects[progobj.subproject].held_object
-                else:
-                    interp = self
-                assert isinstance(interp, Interpreter)
-                version = interp.project_version
-            else:
-                version = progobj.get_version(self)
-            is_found, not_found, _ = mesonlib.version_compare_many(version, wanted)
-            if not is_found:
-                mlog.log('Program', mlog.bold(progobj.name), 'found:', mlog.red('NO'),
-                         'found', mlog.normal_cyan(version), 'but need:',
-                         mlog.bold(', '.join([f"'{e}'" for e in not_found])), *extra_info)
-                if required:
-                    m = 'Invalid version of program, need {!r} {!r} found {!r}.'
-                    raise InterpreterException(m.format(progobj.name, not_found, version))
-                return self.notfound_program(args)
-            extra_info.insert(0, mlog.normal_cyan(version))
+        if not self.check_program_version(progobj, wanted, version_func, extra_info):
+            if required:
+                raise InterpreterException('Invalid version of program.')
```

Comment:
Seems like some of the context got dropped.

`'Invalid version of program, need {!r} {!r} found {!r}.'`

Reply:
check_program_version() already print a complete message, the exception does not need to repeat everything.

Reply:
Logging was indeed a bit broken. It now prints:
```
Program glib-compile-resources found: NO found 2.76.1 but need: '>3.0' (/usr/bin/glib-compile-resources)

meson.build:4:0: ERROR: Program 'glib-compile-resources' not found or not executable
```

And if a fallback is used:
```
Program prog-version.py prog-version found: NO found 1.0 but need: '>= 2.0' (/home/xclaesse/programmation/meson/test cases/common/182 find override/prog-version.py)
Fallback to subproject sub2 which provides program prog-version.py prog-version
...
Program prog-version.py found: YES 2.0 (overridden)
```